Usability/split up pointer primitive checks

hannes-steffenhagen-diffblue opened the pull request:
The pointer-primitive-checks require an address check. Previous we’ve just done these all in in one go, but this makes it hard to see _which_ of our address checks failed.

See https://github.com/diffblue/cbmc/issues/5531
- [ ] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [ ] My PR is restricted to a single feature or bugfix.
- [ ] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

martin-cs commented on regression/cbmc/pointer-primitive-check-01/test.desc:
```diff
@@ -1,15 +1,43 @@
-CORE
+CORE broken-smt-backend
```
Why `broken-smt-backend` when it wasn't before?

hannes-steffenhagen-diffblue replied:
@martin-cs I am not sure _why_ but some individual pointer checks seem to have different results in the smt backend. Because in the previous version of this we AND-ed them all together this didn’t matter before.

martin-cs replied:
@hannes-steffenhagen-diffblue Thanks for the explanation but this does leave me a little concerned.  They really shouldn't do.  This is kind of important functionality.

martin-cs commented on src/analyses/goto_check.cpp:
```diff
@@ -1226,21 +1230,21 @@ void goto_checkt::pointer_primitive_check(
+  {
+    // we want all address checks *except* for the null check because these
+    // should only fail if the assumption is on in invalid non-null pointer.
+    if(c.description != pointer_is_null_check_description)
```
This seems to have less checks?  Thanks for adding the comment but I think I am struggling to understand it, possibly because I am missing the context.  Please can you explain it a bit more to me?

tautschnig replied:
Adding to this: is a string comparison really the best solution? Couldn't we instead compare `c.assertion` to `null_object(pointer)`?

hannes-steffenhagen-diffblue replied:
@tautschnig I don’t like it either, but I don’t like repeating the implementation of `c.assertion` much better. My preferred thing to do would be to add stable IDs to the checks we generate so we can check that, e.g. this is "the address null check", something along those lines. This would be a bigger change though, not sure if people would be OK with that approach?

tautschnig replied:
How about adding further members to `conditiont` that track which kinds of properties are already asserted? We don't have to be complete right away, you could just start with a single one: `bool has_null_check`.

hannes-steffenhagen-diffblue replied:
@tautschnig One option to get this through right now would be to park the idea of how to filter out checks for later and just accept the redundant null-check for now, but that doesn’t make me happy either.

tautschnig replied:
Well, why not try to route of adding, e.g., `has_null_check`?

hannes-steffenhagen-diffblue replied:
@tautschnig I don’t completely understand how you’d want that to work – I want to remove the null check, not ask whether it has one.

tautschnig replied:
I'd basically use a `bool` member instead of the string. But then what @thomasspriggs [said](https://github.com/diffblue/cbmc/pull/5582/files#r525275170) seems a lot better.

tautschnig commented on src/analyses/goto_check.cpp:
```diff
@@ -1226,21 +1230,21 @@ void goto_checkt::pointer_primitive_check(
-    guard);
+  {
+    // we want all address checks *except* for the null check because these
+    // should only fail if the assumption is on in invalid non-null pointer.
```
NIt picking: s/on in/on an/

tautschnig reviewed (commented):
As said in my comment below, I'm not convinced the string comparison is the right approach. Apart from this: Please also add a "Fixes: #5531" line to the commit message.

thomasspriggs commented on src/analyses/goto_check.cpp:
```diff
@@ -1226,21 +1230,21 @@ void goto_checkt::pointer_primitive_check(
+    if(c.description != pointer_is_null_check_description)
+    {
+      add_guarded_property(
+        or_exprt{null_object(pointer), c.assertion},
```
:no_entry_sign: This doesn't actually appear to be adding the equivalent checks in the case where `mode == ID_java` or where `flags.is_null()` is false.

There must be a way to implement this which is both more readable and more straight forward to get correct. Besides the issue of the string comparison not being great, I think that filtering the list of conditions after generating them is fundamentally a sub-optimal approach. I would suggest starting by looking at splitting the `goto_checkt::address_check` function into multiple functions. At least one of which should generate the list of conditions without the null check. That way the check never has to be filtered out after being unnecessarily added.

tautschnig commented:
@hannes-steffenhagen-diffblue Please let us know when this is ready for review - I like the direction this is going, but 1) some cleanup of the previous approach is needed and 2) tests seem to be failing.

thomasspriggs commented on src/analyses/goto_check.cpp:
```diff
@@ -39,6 +39,8 @@ Author: Daniel Kroening, kroening@kroening.com
 #include "guard.h"
 #include "local_bitvector_analysis.h"
 
+static constexpr const char *pointer_is_null_check_description = "pointer NULL";
```
:no_entry_sign: Adding this when we don't need it is now unnecessary churn. It should be removed so that we are treating all the descriptions in the same way.

thomasspriggs commented on src/analyses/goto_check.cpp:
```diff
@@ -164,6 +166,8 @@ class goto_checkt
     }
 
     exprt assertion;
+    // pointer-primitive-check assumes that these are fixed. If we ever add I18N
```
:question: What do you mean by "fixed"? Is this an assumption that is still being made with the current refactor? Is `pointer_validity_check` making the same assumption?

thomasspriggs replied:
:pick: I am not sure the comment about internationalisation is useful. Surely that would affect strings throughout the code base, not just this one. That also seems like the implementation choices for that are something which would be decided when we get to it rather than something which is a useful comment.

thomasspriggs commented on src/analyses/goto_check.cpp:
```diff
@@ -2228,6 +2133,134 @@ void goto_checkt::goto_check(
+  const exprt &size)
+{
+  PRECONDITION(local_bitvector_analysis);
+  PRECONDITION(address.type().id() == ID_pointer);
```
:bulb: You could just do `const auto &pointer_type = type_checked_cast<pointer_typet>(address.type());` rather than duplicating the same check inside `to_pointer_type` and taking 2 lines to do this.

thomasspriggs commented on src/analyses/goto_check.cpp:
```diff
@@ -202,7 +206,18 @@ class goto_checkt
 
-  conditionst address_check(const exprt &address, const exprt &size);
+  optionalt<goto_checkt::conditiont>
+  get_pointer_is_null_condition(const exprt &address, const exprt &size);
```
:pick: It would be nice to have some doxygen for these functions.

thomasspriggs reviewed (commented):
This is looking better. I like that we no longer need the filtering. :+1: . However I think this needs to be split into more commits before merging (and maybe before fixing the tests). This is because we are now doing multiple distinct different things within the same commit. Keeping them in separate commits would help with both debugging the tests right now and making the commit history make sense which could help with tracking down future issues. I think I would make the commit history look something like this -

1. Reorder existing checks in `address_check`. As in - put the null check first only. The tests can then be updated for this functionality only which will help with testing the next commit.
2. Refactor to split the `address_check` function. This commit should include no functionality changes. That way we can use the tests from commit 1 to test that this refactor is sound.
3. Separate calls to `add_guarded_property` in `pointer_primitive_check`. Leave the null condition in place.
4. Remove the null condition in `pointer_primitive_check`. I am assuming you did this removal in order to avoid adding a condition which would otherwise be checking "x is null or x is not null". However the commit history does not currently make this clear and putting this in a separate commit will allow for putting the motivation in the commit message body.

I am a little on the fence about the ordering 3/4.

hannes-steffenhagen-diffblue committed:
Reorder address checks

This is a slight change in functionality. A future refactor of the
address checks will necessitate reordering these anyway, and doing it in
isolation first gives us some additional confidence that our refactor
didn’t have any unintended consequences aside from the reordering.

Also changes the regression tests that break because of the reorder.

hannes-steffenhagen-diffblue committed:
Refactor address_check

1. Renames address_check to get_pointer_dereferenceable_conditions.
   The reasoning here is that the various _check functions should
   correspond to the --*-check options, whereas this function is
   a helper used by these
2. Split its implementation to separate the null check (which is done
   differently in Java than in C or languages with C-like semantics) and
   the other checks (which are not to be done in Java at all).

hannes-steffenhagen-diffblue committed:
Split up pointer-primitive-checks

These were previously checking if the pointer being validated is
either null or a valid pointer, because primitives like r_ok don't work
on invalid pointers. However, grouping them all together like that makes
it hard to see which exact condition failed. This just adds each of
these checks individually instead, which is more verbose but therefore
also more useful when diagnosing problems.

Some of the tests are marked as broken-smt-backend because the SMT
backend gives different results for these checks for some reason.

thomasspriggs reviewed (approved):
Thank you for putting the work into refactoring this and cleaning up the commit history. This version was _much_ easier to review and this implementation should be easier to maintain.

chrisr-diffblue reviewed (approved):
Generally looks good, thanks - I'd recommend getting a re-review from @tautschnig too, just to check he's happy with the updates.

hannes-steffenhagen-diffblue commented:
@tautschnig are you OK with this now?
